```
{ ScriptError 
level=warning msg="[linters_context] gocritic: no need to enable check \"deprecatedComment\": it's already enabled"
cmd/kubelet/app/server.go:98:2: could not import k8s.io/kubernetes/pkg/kubelet (-: # k8s.io/kubernetes/pkg/kubelet
pkg/kubelet/kubelet.go:1706:6: no new variables on left side of :=) (typecheck)
	"k8s.io/kubernetes/pkg/kubelet"
	^
pkg/kubelet/active_deadline.go:1: : # k8s.io/kubernetes/pkg/kubelet [k8s.io/kubernetes/pkg/kubelet.test]
pkg/kubelet/kubelet.go:1706:6: no new variables on left side of := (typecheck)
/*
pkg/kubemark/hollow_kubelet.go:35:2: could not import k8s.io/kubernetes/pkg/kubelet (-: # k8s.io/kubernetes/pkg/kubelet
pkg/kubelet/kubelet.go:1706:6: no new variables on left side of :=) (typecheck)
	"k8s.io/kubernetes/pkg/kubelet"
	^
test/e2e/common/node/pods.go:51:2: could not import k8s.io/kubernetes/pkg/kubelet (-: # k8s.io/kubernetes/pkg/kubelet
pkg/kubelet/kubelet.go:1706:6: no new variables on left side of :=) (typecheck)
	"k8s.io/kubernetes/pkg/kubelet"
	^
test/e2e_node/log_path_test.go:25:2: could not import k8s.io/kubernetes/pkg/kubelet (-: # k8s.io/kubernetes/pkg/kubelet
pkg/kubelet/kubelet.go:1706:6: no new variables on left side of :=) (typecheck)
	"k8s.io/kubernetes/pkg/kubelet"
	^
```

@bart0sh would you rebase with latest master branch? There seems to be some changes conflicts.

--- a/hack/golangci-hints.yaml
+++ b/hack/golangci-hints.yaml
@@ -235,6 +235,7 @@ linters:
             contextual k8s.io/kubernetes/pkg/kubelet/preemption/.*
             contextual k8s.io/kubernetes/pkg/kubelet/volumemanager/.*
             contextual k8s.io/kubernetes/pkg/kubelet/util/.*
+            contextual k8s.io/kubernetes/pkg/kubelet/logs/.*
             
             # As long as contextual logging is alpha or beta, all WithName, WithValues,
             # NewContext calls have to go through klog. Once it is GA, we can lift

--- a/hack/golangci-hints.yaml
+++ b/hack/golangci-hints.yaml
@@ -234,6 +234,7 @@ linters:
             contextual k8s.io/kubernetes/pkg/kubelet/pod/.*
             contextual k8s.io/kubernetes/pkg/kubelet/preemption/.*
             contextual k8s.io/kubernetes/pkg/kubelet/volumemanager/.*
+            contextual k8s.io/kubernetes/pkg/kubelet/util/.*
             
             # As long as contextual logging is alpha or beta, all WithName, WithValues,
             # NewContext calls have to go through klog. Once it is GA, we can lift

--- a/hack/golangci.yaml
+++ b/hack/golangci.yaml
@@ -249,6 +249,7 @@ linters:
             contextual k8s.io/kubernetes/pkg/kubelet/preemption/.*
             contextual k8s.io/kubernetes/pkg/kubelet/volumemanager/.*
             contextual k8s.io/kubernetes/pkg/kubelet/util/.*
+            contextual k8s.io/kubernetes/pkg/kubelet/logs/.*
             
             # As long as contextual logging is alpha or beta, all WithName, WithValues,
             # NewContext calls have to go through klog. Once it is GA, we can lift

--- a/hack/golangci.yaml
+++ b/hack/golangci.yaml
@@ -248,6 +248,7 @@ linters:
             contextual k8s.io/kubernetes/pkg/kubelet/pod/.*
             contextual k8s.io/kubernetes/pkg/kubelet/preemption/.*
             contextual k8s.io/kubernetes/pkg/kubelet/volumemanager/.*
+            contextual k8s.io/kubernetes/pkg/kubelet/util/.*
             
             # As long as contextual logging is alpha or beta, all WithName, WithValues,
             # NewContext calls have to go through klog. Once it is GA, we can lift

--- a/hack/logcheck.conf
+++ b/hack/logcheck.conf
@@ -80,6 +80,7 @@ contextual k8s.io/kubernetes/pkg/kubelet/nodeshutdown/.*
 contextual k8s.io/kubernetes/pkg/kubelet/pod/.*
 contextual k8s.io/kubernetes/pkg/kubelet/preemption/.*
 contextual k8s.io/kubernetes/pkg/kubelet/volumemanager/.*
+contextual k8s.io/kubernetes/pkg/kubelet/util/.*
 
 # As long as contextual logging is alpha or beta, all WithName, WithValues,
 # NewContext calls have to go through klog. Once it is GA, we can lift

--- a/hack/logcheck.conf
+++ b/hack/logcheck.conf
@@ -81,6 +81,7 @@ contextual k8s.io/kubernetes/pkg/kubelet/pod/.*
 contextual k8s.io/kubernetes/pkg/kubelet/preemption/.*
 contextual k8s.io/kubernetes/pkg/kubelet/volumemanager/.*
 contextual k8s.io/kubernetes/pkg/kubelet/util/.*
+contextual k8s.io/kubernetes/pkg/kubelet/logs/.*
 
 # As long as contextual logging is alpha or beta, all WithName, WithValues,
 # NewContext calls have to go through klog. Once it is GA, we can lift

--- a/pkg/kubelet/kubelet.go
+++ b/pkg/kubelet/kubelet.go
@@ -1715,7 +1715,7 @@ func (kl *Kubelet) initializeRuntimeDependentModules() {
 
 	// container log manager must start after container runtime is up to retrieve information from container runtime
 	// and inform container to reopen log file after log rotation.
-	kl.containerLogManager.Start()
+	kl.containerLogManager.Start(ctx)
 	// Adding Registration Callback function for CSI Driver
 	kl.pluginManager.AddHandler(pluginwatcherapi.CSIPlugin, plugincache.PluginHandler(csi.PluginHandler))
 	// Adding Registration Callback function for DRA Plugin and Device Plugin

--- a/pkg/kubelet/kubelet.go
+++ b/pkg/kubelet/kubelet.go
@@ -1059,7 +1059,7 @@ func NewMainKubelet(ctx context.Context,
 		renewInterval,
 		string(klet.nodeName),
 		v1.NamespaceNodeLease,
-		util.SetNodeOwnerFunc(klet.heartbeatClient, string(klet.nodeName)))
+		util.SetNodeOwnerFunc(ctx, klet.heartbeatClient, string(klet.nodeName)))
 
 	// setup node shutdown manager
 	shutdownManager := nodeshutdown.NewManager(&nodeshutdown.Config{

--- a/pkg/kubelet/logs/container_log_manager.go
+++ b/pkg/kubelet/logs/container_log_manager.go
@@ -54,7 +54,7 @@ const (
 // Implementation is thread-safe.
 type ContainerLogManager interface {
 	// Start container log manager.
-	Start()
+	Start(ctx context.Context)
 	// Clean removes all logs of specified container.
 	Clean(ctx context.Context, containerID string) error
 }
@@ -141,9 +141,9 @@ func NewContainerLogManager(runtimeService internalapi.RuntimeService, osInterfa
 }
 
 // Start the container log manager.
-func (c *containerLogManager) Start() {
-	ctx := context.Background()
-	klog.InfoS("Initializing container log rotate workers", "workers", c.maxWorkers, "monitorPeriod", c.monitoringPeriod)
+func (c *containerLogManager) Start(ctx context.Context) {
+	logger := klog.FromContext(ctx)
+	logger.Info("Initializing container log rotate workers", "workers", c.maxWorkers, "monitorPeriod", c.monitoringPeriod)
 	for i := 0; i < c.maxWorkers; i++ {
 		worker := i + 1
 		go c.processQueueItems(ctx, worker)
@@ -151,7 +151,7 @@ func (c *containerLogManager) Start() {
 	// Start a goroutine periodically does container log rotation.
 	go wait.Forever(func() {
 		if err := c.rotateLogs(ctx); err != nil {
-			klog.ErrorS(err, "Failed to rotate container logs")
+			logger.Error(err, "Failed to rotate container logs")
 		}
 	}, c.monitoringPeriod.Duration)
 }
@@ -183,16 +183,19 @@ func (c *containerLogManager) Clean(ctx context.Context, containerID string) err
 }
 
 func (c *containerLogManager) processQueueItems(ctx context.Context, worker int) {
-	klog.V(4).InfoS("Starting container log rotation worker", "workerID", worker)
+	logger := klog.FromContext(ctx)
+	logger.V(4).Info("Starting container log rotation worker", "workerID", worker)
 	for c.processContainer(ctx, worker) {
 	}
-	klog.V(4).InfoS("Terminating container log rotation worker", "workerID", worker)
+	logger.V(4).Info("Terminating container log rotation worker", "workerID", worker)
 }
 
 func (c *containerLogManager) rotateLogs(ctx context.Context) error {
+	logger := klog.FromContext(ctx)
 	c.mutex.Lock()
 	defer c.mutex.Unlock()
-	klog.V(4).InfoS("Starting container log rotation sequence")
+	logger.V(4).Info("Starting container log rotation sequence")
+	// TODO(#59998): Use kubelet pod cache.
 	containers, err := c.runtimeService.ListContainers(ctx, &runtimeapi.ContainerFilter{})
 	if err != nil {
 		return fmt.Errorf("failed to list containers: %w", err)
@@ -204,8 +207,8 @@ func (c *containerLogManager) rotateLogs(ctx context.Context) error {
 			continue
 		}
 		// Doing this to avoid additional overhead with logging of label like arguments that can prove costly
-		if v := klog.V(4); v.Enabled() {
-			klog.V(4).InfoS("Adding new entry to the queue for processing", "id", container.GetId(), "name", container.Metadata.GetName(), "labels", container.GetLabels())
+		if v := logger.V(4); v.Enabled() {
+			logger.V(4).Info("Adding new entry to the queue for processing", "id", container.GetId(), "name", container.Metadata.GetName(), "labels", container.GetLabels())
 		}
 		c.queue.Add(container.GetId())
 	}
@@ -224,14 +227,14 @@ func (c *containerLogManager) processContainer(ctx context.Context, worker int) 
 	// Always default the return to true to keep the processing of Queue ongoing
 	ok = true
 	id := key
-
+	logger := klog.FromContext(ctx)
 	resp, err := c.runtimeService.ContainerStatus(ctx, id, false)
 	if err != nil {
-		klog.ErrorS(err, "Failed to get container status", "worker", worker, "containerID", id)
+		logger.Error(err, "Failed to get container status", "worker", worker, "containerID", id)
 		return
 	}
 	if resp.GetStatus() == nil {
-		klog.ErrorS(err, "Container status is nil", "worker", worker, "containerID", id)
+		logger.Error(err, "Container status is nil", "worker", worker, "containerID", id)
 		return
 	}
 	path := resp.GetStatus().GetLogPath()
@@ -239,28 +242,28 @@ func (c *containerLogManager) processContainer(ctx context.Context, worker int) 
 
 	if err != nil {
 		if !os.IsNotExist(err) {
-			klog.ErrorS(err, "Failed to stat container log", "worker", worker, "containerID", id, "path", path)
+			logger.Error(err, "Failed to stat container log", "worker", worker, "containerID", id, "path", path)
 			return
 		}
 
 		if err = c.runtimeService.ReopenContainerLog(ctx, id); err != nil {
-			klog.ErrorS(err, "Container log doesn't exist, reopen container log failed", "worker", worker, "containerID", id, "path", path)
+			logger.Error(err, "Container log doesn't exist, reopen container log failed", "worker", worker, "containerID", id, "path", path)
 			return
 		}
 
 		info, err = c.osInterface.Stat(path)
 		if err != nil {
-			klog.ErrorS(err, "Failed to stat container log after reopen", "worker", worker, "containerID", id, "path", path)
+			logger.Error(err, "Failed to stat container log after reopen", "worker", worker, "containerID", id, "path", path)
 			return
 		}
 	}
 	if info.Size() < c.policy.MaxSize {
-		klog.V(7).InfoS("log file doesn't need to be rotated", "worker", worker, "containerID", id, "path", path, "currentSize", info.Size(), "maxSize", c.policy.MaxSize)
+		logger.V(7).Info("log file doesn't need to be rotated", "worker", worker, "containerID", id, "path", path, "currentSize", info.Size(), "maxSize", c.policy.MaxSize)
 		return
 	}
 
 	if err := c.rotateLog(ctx, id, path); err != nil {
-		klog.ErrorS(err, "Failed to rotate log for container", "worker", worker, "containerID", id, "path", path, "currentSize", info.Size(), "maxSize", c.policy.MaxSize)
+		logger.Error(err, "Failed to rotate log for container", "worker", worker, "containerID", id, "path", path, "currentSize", info.Size(), "maxSize", c.policy.MaxSize)
 		return
 	}
 	return
@@ -411,6 +414,7 @@ func (c *containerLogManager) compressLog(log string) error {
 // rotateLatestLog rotates latest log without compression, so that container can still write
 // and fluentd can finish reading.
 func (c *containerLogManager) rotateLatestLog(ctx context.Context, id, log string) error {
+	logger := klog.FromContext(ctx)
 	timestamp := c.clock.Now().Format(timestampFormat)
 	rotated := fmt.Sprintf("%s.%s", log, timestamp)
 	if err := c.osInterface.Rename(log, rotated); err != nil {
@@ -424,7 +428,7 @@ func (c *containerLogManager) rotateLatestLog(ctx context.Context, id, log strin
 			// This shouldn't happen.
 			// Report an error if this happens, because we will lose original
 			// log.
-			klog.ErrorS(renameErr, "Failed to rename rotated log", "rotatedLog", rotated, "newLog", log, "containerID", id)
+			logger.Error(renameErr, "Failed to rename rotated log", "rotatedLog", rotated, "newLog", log, "containerID", id)
 		}
 		return fmt.Errorf("failed to reopen container log %q: %w", id, err)
 	}

--- a/pkg/kubelet/logs/container_log_manager_stub.go
+++ b/pkg/kubelet/logs/container_log_manager_stub.go
@@ -20,7 +20,7 @@ import "context"
 
 type containerLogManagerStub struct{}
 
-func (*containerLogManagerStub) Start() {}
+func (*containerLogManagerStub) Start(ctx context.Context) {}
 
 func (*containerLogManagerStub) Clean(ctx context.Context, containerID string) error {
 	return nil

--- a/pkg/kubelet/logs/container_log_manager_test.go
+++ b/pkg/kubelet/logs/container_log_manager_test.go
@@ -31,6 +31,7 @@ import (
 	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/client-go/util/workqueue"
 	"k8s.io/kubernetes/pkg/kubelet/container"
+	"k8s.io/kubernetes/test/utils/ktesting"
 
 	runtimeapi "k8s.io/cri-api/pkg/apis/runtime/v1"
 	critest "k8s.io/cri-api/pkg/apis/testing"
@@ -77,7 +78,7 @@ func TestGetAllLogs(t *testing.T) {
 }
 
 func TestRotateLogs(t *testing.T) {
-	ctx := context.Background()
+	tCtx := ktesting.Init(t)
 	dir, err := os.MkdirTemp("", "test-rotate-logs")
 	require.NoError(t, err)
 	defer os.RemoveAll(dir)
@@ -160,12 +161,12 @@ func TestRotateLogs(t *testing.T) {
 	f.SetFakeContainers(testContainers)
 
 	// Push the items into the queue for before starting the worker to avoid issue with the queue being empty.
-	require.NoError(t, c.rotateLogs(ctx))
+	require.NoError(t, c.rotateLogs(tCtx))
 
 	// Start a routine that can monitor the queue and shutdown the queue to trigger the retrun from the processQueueItems
 	// Keeping the monitor duration smaller in order to keep the unwanted delay in the test to a minimal.
 	go func() {
-		pollTimeoutCtx, cancel := context.WithTimeout(ctx, 10*time.Second)
+		pollTimeoutCtx, cancel := context.WithTimeout(tCtx, 10*time.Second)
 		defer cancel()
 		err = wait.PollUntilContextCancel(pollTimeoutCtx, 5*time.Millisecond, false, func(ctx context.Context) (done bool, err error) {
 			return c.queue.Len() == 0, nil
@@ -176,7 +177,7 @@ func TestRotateLogs(t *testing.T) {
 		c.queue.ShutDown()
 	}()
 	// This is a blocking call. But the above routine takes care of ensuring that this is terminated once the queue is shutdown
-	c.processQueueItems(ctx, 1)
+	c.processQueueItems(tCtx, 1)
 
 	timestamp := now.Format(timestampFormat)
 	logs, err := os.ReadDir(dir)
@@ -190,7 +191,7 @@ func TestRotateLogs(t *testing.T) {
 }
 
 func TestClean(t *testing.T) {
-	ctx := context.Background()
+	tCtx := ktesting.Init(t)
 	dir, err := os.MkdirTemp("", "test-clean")
 	require.NoError(t, err)
 	defer os.RemoveAll(dir)
@@ -256,7 +257,7 @@ func TestClean(t *testing.T) {
 	}
 	f.SetFakeContainers(testContainers)
 
-	err = c.Clean(ctx, "container-3")
+	err = c.Clean(tCtx, "container-3")
 	require.NoError(t, err)
 
 	logs, err := os.ReadDir(dir)
@@ -387,7 +388,7 @@ func TestCompressLog(t *testing.T) {
 }
 
 func TestRotateLatestLog(t *testing.T) {
-	ctx := context.Background()
+	tCtx := ktesting.Init(t)
 	dir, err := os.MkdirTemp("", "test-rotate-latest-log")
 	require.NoError(t, err)
 	defer os.RemoveAll(dir)
@@ -438,7 +439,7 @@ func TestRotateLatestLog(t *testing.T) {
 		defer testFile.Close()
 		testLog := testFile.Name()
 		rotatedLog := fmt.Sprintf("%s.%s", testLog, now.Format(timestampFormat))
-		err = c.rotateLatestLog(ctx, "test-id", testLog)
+		err = c.rotateLatestLog(tCtx, "test-id", testLog)
 		assert.Equal(t, test.expectError, err != nil)
 		_, err = os.Stat(testLog)
 		assert.Equal(t, test.expectOriginal, err == nil)

--- a/pkg/kubelet/util/boottime_util_linux.go
+++ b/pkg/kubelet/util/boottime_util_linux.go
@@ -20,6 +20,7 @@ limitations under the License.
 package util
 
 import (
+	"context"
 	"fmt"
 	"os"
 	"strconv"
@@ -36,7 +37,10 @@ import (
 func GetBootTime() (time.Time, error) {
 	bootTime, err := getBootTimeWithProcStat()
 	if err != nil {
-		klog.InfoS("Failed to get boot time from /proc/uptime. Will retry with unix.Sysinfo.", "error", err)
+		// TODO: it needs to be replaced by a proper context in the future
+		ctx := context.TODO()
+		logger := klog.FromContext(ctx)
+		logger.Info("Failed to get boot time from /proc/uptime. Will retry with unix.Sysinfo.", "error", err)
 		return getBootTimeWithSysinfo()
 	}
 	return bootTime, nil

--- a/pkg/kubelet/util/manager/cache_based_manager_test.go
+++ b/pkg/kubelet/util/manager/cache_based_manager_test.go
@@ -37,6 +37,7 @@ import (
 	"k8s.io/client-go/kubernetes/fake"
 	core "k8s.io/client-go/testing"
 	podutil "k8s.io/kubernetes/pkg/api/v1/pod"
+	"k8s.io/kubernetes/test/utils/ktesting"
 	"k8s.io/utils/clock"
 	testingclock "k8s.io/utils/clock/testing"
 
@@ -57,15 +58,15 @@ func noObjectTTL() (time.Duration, bool) {
 	return time.Duration(0), false
 }
 
-func getSecret(fakeClient clientset.Interface) GetObjectFunc {
+func getSecret(ctx context.Context, fakeClient clientset.Interface) GetObjectFunc {
 	return func(namespace, name string, opts metav1.GetOptions) (runtime.Object, error) {
-		return fakeClient.CoreV1().Secrets(namespace).Get(context.TODO(), name, opts)
+		return fakeClient.CoreV1().Secrets(namespace).Get(ctx, name, opts)
 	}
 }
 
-func newSecretStore(fakeClient clientset.Interface, clock clock.Clock, getTTL GetObjectTTLFunc, ttl time.Duration) *objectStore {
+func newSecretStore(ctx context.Context, fakeClient clientset.Interface, clock clock.Clock, getTTL GetObjectTTLFunc, ttl time.Duration) *objectStore {
 	return &objectStore{
-		getObject:  getSecret(fakeClient),
+		getObject:  getSecret(ctx, fakeClient),
 		clock:      clock,
 		items:      make(map[objectKey]*objectStoreItem),
 		defaultTTL: ttl,
@@ -87,8 +88,9 @@ func newCacheBasedSecretManager(store Store) Manager {
 }
 
 func TestSecretStore(t *testing.T) {
+	tCtx := ktesting.Init(t)
 	fakeClient := &fake.Clientset{}
-	store := newSecretStore(fakeClient, clock.RealClock{}, noObjectTTL, 0)
+	store := newSecretStore(tCtx, fakeClient, clock.RealClock{}, noObjectTTL, 0)
 	store.AddReference("ns1", "name1", "pod1")
 	store.AddReference("ns2", "name2", "pod2")
 	store.AddReference("ns1", "name1", "pod3")
@@ -121,8 +123,9 @@ func TestSecretStore(t *testing.T) {
 }
 
 func TestSecretStoreDeletingSecret(t *testing.T) {
+	tCtx := ktesting.Init(t)
 	fakeClient := &fake.Clientset{}
-	store := newSecretStore(fakeClient, clock.RealClock{}, noObjectTTL, 0)
+	store := newSecretStore(tCtx, fakeClient, clock.RealClock{}, noObjectTTL, 0)
 	store.AddReference("ns", "name", "pod")
 
 	result := &v1.Secret{ObjectMeta: metav1.ObjectMeta{Namespace: "ns", Name: "name", ResourceVersion: "10"}}
@@ -150,9 +153,10 @@ func TestSecretStoreDeletingSecret(t *testing.T) {
 }
 
 func TestSecretStoreGetAlwaysRefresh(t *testing.T) {
+	tCtx := ktesting.Init(t)
 	fakeClient := &fake.Clientset{}
 	fakeClock := testingclock.NewFakeClock(time.Now())
-	store := newSecretStore(fakeClient, fakeClock, noObjectTTL, 0)
+	store := newSecretStore(tCtx, fakeClient, fakeClock, noObjectTTL, 0)
 
 	for i := 0; i < 10; i++ {
 		store.AddReference(fmt.Sprintf("ns-%d", i), fmt.Sprintf("name-%d", i), types.UID(fmt.Sprintf("pod-%d", i)))
@@ -177,9 +181,10 @@ func TestSecretStoreGetAlwaysRefresh(t *testing.T) {
 }
 
 func TestSecretStoreGetNeverRefresh(t *testing.T) {
+	tCtx := ktesting.Init(t)
 	fakeClient := &fake.Clientset{}
 	fakeClock := testingclock.NewFakeClock(time.Now())
-	store := newSecretStore(fakeClient, fakeClock, noObjectTTL, time.Minute)
+	store := newSecretStore(tCtx, fakeClient, fakeClock, noObjectTTL, time.Minute)
 
 	for i := 0; i < 10; i++ {
 		store.AddReference(fmt.Sprintf("ns-%d", i), fmt.Sprintf("name-%d", i), types.UID(fmt.Sprintf("pod-%d", i)))
@@ -207,9 +212,10 @@ func TestCustomTTL(t *testing.T) {
 		return ttl, ttlExists
 	}
 
+	tCtx := ktesting.Init(t)
 	fakeClient := &fake.Clientset{}
 	fakeClock := testingclock.NewFakeClock(time.Time{})
-	store := newSecretStore(fakeClient, fakeClock, customTTL, time.Minute)
+	store := newSecretStore(tCtx, fakeClient, fakeClock, customTTL, time.Minute)
 
 	store.AddReference("ns", "name", "pod")
 	store.Get("ns", "name")
@@ -383,9 +389,10 @@ func podWithSecretsAndUID(ns, podName, podUID string, toAttach secretsToAttach) 
 }
 
 func TestCacheInvalidation(t *testing.T) {
+	tCtx := ktesting.Init(t)
 	fakeClient := &fake.Clientset{}
 	fakeClock := testingclock.NewFakeClock(time.Now())
-	store := newSecretStore(fakeClient, fakeClock, noObjectTTL, time.Minute)
+	store := newSecretStore(tCtx, fakeClient, fakeClock, noObjectTTL, time.Minute)
 	manager := newCacheBasedSecretManager(store)
 
 	// Create a pod with some secrets.
@@ -438,9 +445,10 @@ func TestCacheInvalidation(t *testing.T) {
 }
 
 func TestResourceContentExpired(t *testing.T) {
+	tCtx := ktesting.Init(t)
 	fakeClient := &fake.Clientset{}
 	fakeClock := testingclock.NewFakeClock(time.Now())
-	store := newSecretStore(fakeClient, fakeClock, noObjectTTL, time.Minute)
+	store := newSecretStore(tCtx, fakeClient, fakeClock, noObjectTTL, time.Minute)
 	manager := newCacheBasedSecretManager(store)
 
 	// Create a pod with some secrets.
@@ -463,9 +471,10 @@ func TestResourceContentExpired(t *testing.T) {
 }
 
 func TestRegisterIdempotence(t *testing.T) {
+	tCtx := ktesting.Init(t)
 	fakeClient := &fake.Clientset{}
 	fakeClock := testingclock.NewFakeClock(time.Now())
-	store := newSecretStore(fakeClient, fakeClock, noObjectTTL, time.Minute)
+	store := newSecretStore(tCtx, fakeClient, fakeClock, noObjectTTL, time.Minute)
 	manager := newCacheBasedSecretManager(store)
 
 	s1 := secretsToAttach{
@@ -498,9 +507,10 @@ func TestRegisterIdempotence(t *testing.T) {
 }
 
 func TestCacheRefcounts(t *testing.T) {
+	tCtx := ktesting.Init(t)
 	fakeClient := &fake.Clientset{}
 	fakeClock := testingclock.NewFakeClock(time.Now())
-	store := newSecretStore(fakeClient, fakeClock, noObjectTTL, time.Minute)
+	store := newSecretStore(tCtx, fakeClient, fakeClock, noObjectTTL, time.Minute)
 	manager := newCacheBasedSecretManager(store)
 
 	s1 := secretsToAttach{
@@ -619,8 +629,9 @@ func TestCacheRefcounts(t *testing.T) {
 }
 
 func TestCacheBasedSecretManager(t *testing.T) {
+	tCtx := ktesting.Init(t)
 	fakeClient := &fake.Clientset{}
-	store := newSecretStore(fakeClient, clock.RealClock{}, noObjectTTL, 0)
+	store := newSecretStore(tCtx, fakeClient, clock.RealClock{}, noObjectTTL, 0)
 	manager := newCacheBasedSecretManager(store)
 
 	// Create a pod with some secrets.

--- a/pkg/kubelet/util/manager/watch_based_manager.go
+++ b/pkg/kubelet/util/manager/watch_based_manager.go
@@ -17,6 +17,7 @@ limitations under the License.
 package manager
 
 import (
+	"context"
 	"fmt"
 	"sync"
 	"time"
@@ -343,7 +344,10 @@ func (c *objectCache) Get(namespace, name string) (runtime.Object, error) {
 		if c.isImmutable(object) {
 			item.setImmutable()
 			if item.stop() {
-				klog.V(4).InfoS("Stopped watching for changes - object is immutable", "obj", klog.KRef(namespace, name))
+				// TODO: it needs to be replaced by a proper context in the future
+				ctx := context.TODO()
+				logger := klog.FromContext(ctx)
+				logger.V(4).Info("Stopped watching for changes - object is immutable", "obj", klog.KRef(namespace, name))
 			}
 		}
 		return object, nil
@@ -352,12 +356,15 @@ func (c *objectCache) Get(namespace, name string) (runtime.Object, error) {
 }
 
 func (c *objectCache) startRecycleIdleWatch() {
+	// TODO: it needs to be replaced by a proper context in the future
+	ctx := context.TODO()
+	logger := klog.FromContext(ctx)
 	c.lock.Lock()
 	defer c.lock.Unlock()
 
 	for key, item := range c.items {
 		if item.stopIfIdle(c.clock.Now(), c.maxIdleTime) {
-			klog.V(4).InfoS("Not acquired for long time, Stopped watching for changes", "objectKey", key, "maxIdleTime", c.maxIdleTime)
+			logger.V(4).Info("Not acquired for long time, Stopped watching for changes", "objectKey", key, "maxIdleTime", c.maxIdleTime)
 		}
 	}
 }

--- a/pkg/kubelet/util/manager/watch_based_manager_test.go
+++ b/pkg/kubelet/util/manager/watch_based_manager_test.go
@@ -45,15 +45,15 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
-func listSecret(fakeClient clientset.Interface) listObjectFunc {
+func listSecret(ctx context.Context, fakeClient clientset.Interface) listObjectFunc {
 	return func(namespace string, opts metav1.ListOptions) (runtime.Object, error) {
-		return fakeClient.CoreV1().Secrets(namespace).List(context.TODO(), opts)
+		return fakeClient.CoreV1().Secrets(namespace).List(ctx, opts)
 	}
 }
 
-func watchSecret(fakeClient clientset.Interface) watchObjectFunc {
+func watchSecret(ctx context.Context, fakeClient clientset.Interface) watchObjectFunc {
 	return func(namespace string, opts metav1.ListOptions) (watch.Interface, error) {
-		return fakeClient.CoreV1().Secrets(namespace).Watch(context.TODO(), opts)
+		return fakeClient.CoreV1().Secrets(namespace).Watch(ctx, opts)
 	}
 }
 
@@ -64,10 +64,10 @@ func isSecretImmutable(object runtime.Object) bool {
 	return false
 }
 
-func newSecretCache(fakeClient clientset.Interface, fakeClock clock.Clock, maxIdleTime time.Duration) *objectCache {
+func newSecretCache(ctx context.Context, fakeClient clientset.Interface, fakeClock clock.Clock, maxIdleTime time.Duration) *objectCache {
 	return &objectCache{
-		listObject:    listSecret(fakeClient),
-		watchObject:   watchSecret(fakeClient),
+		listObject:    listSecret(ctx, fakeClient),
+		watchObject:   watchSecret(ctx, fakeClient),
 		newObject:     func() runtime.Object { return &v1.Secret{} },
 		isImmutable:   isSecretImmutable,
 		groupResource: corev1.Resource("secret"),
@@ -78,6 +78,7 @@ func newSecretCache(fakeClient clientset.Interface, fakeClock clock.Clock, maxId
 }
 
 func TestSecretCache(t *testing.T) {
+	tCtx := ktesting.Init(t)
 	fakeClient := &fake.Clientset{}
 
 	listReactor := func(a core.Action) (bool, runtime.Object, error) {
@@ -93,7 +94,7 @@ func TestSecretCache(t *testing.T) {
 	fakeClient.AddWatchReactor("secrets", core.DefaultWatchReactor(fakeWatch, nil))
 
 	fakeClock := testingclock.NewFakeClock(time.Now())
-	store := newSecretCache(fakeClient, fakeClock, time.Minute)
+	store := newSecretCache(tCtx, fakeClient, fakeClock, time.Minute)
 
 	store.AddReference("ns", "name", "pod")
 	_, err := store.Get("ns", "name")
@@ -121,7 +122,6 @@ func TestSecretCache(t *testing.T) {
 		return true, nil
 	}
 
-	tCtx := ktesting.Init(t)
 	if err := wait.PollUntilContextCancel(tCtx, 10*time.Millisecond, true, getFn); err != nil {
 		t.Errorf("unexpected error: %v", err)
 	}
@@ -152,6 +152,7 @@ func TestSecretCache(t *testing.T) {
 }
 
 func TestSecretCacheMultipleRegistrations(t *testing.T) {
+	tCtx := ktesting.Init(t)
 	fakeClient := &fake.Clientset{}
 
 	listReactor := func(a core.Action) (bool, runtime.Object, error) {
@@ -167,7 +168,7 @@ func TestSecretCacheMultipleRegistrations(t *testing.T) {
 	fakeClient.AddWatchReactor("secrets", core.DefaultWatchReactor(fakeWatch, nil))
 
 	fakeClock := testingclock.NewFakeClock(time.Now())
-	store := newSecretCache(fakeClient, fakeClock, time.Minute)
+	store := newSecretCache(tCtx, fakeClient, fakeClock, time.Minute)
 
 	store.AddReference("ns", "name", "pod")
 	// This should trigger List and Watch actions eventually.
@@ -184,7 +185,6 @@ func TestSecretCacheMultipleRegistrations(t *testing.T) {
 		}
 		return true, nil
 	}
-	tCtx := ktesting.Init(t)
 	if err := wait.PollUntilContextCancel(tCtx, 10*time.Millisecond, true, actionsFn); err != nil {
 		t.Errorf("unexpected error: %v", err)
 	}
@@ -257,6 +257,7 @@ func TestImmutableSecretStopsTheReflector(t *testing.T) {
 
 	for _, tc := range tests {
 		t.Run(tc.desc, func(t *testing.T) {
+			tCtx := ktesting.Init(t)
 			fakeClient := &fake.Clientset{}
 			listReactor := func(a core.Action) (bool, runtime.Object, error) {
 				result := &v1.SecretList{
@@ -295,7 +296,6 @@ func TestImmutableSecretStopsTheReflector(t *testing.T) {
 
 			// AddReference should start reflector.
 			store.AddReference("ns", "name", "pod")
-			tCtx := ktesting.Init(t)
 			if err := wait.PollUntilContextCancel(tCtx, 10*time.Millisecond, false, itemExists); err != nil {
 				t.Errorf("item wasn't added to cache")
 			}
@@ -340,6 +340,7 @@ func TestImmutableSecretStopsTheReflector(t *testing.T) {
 }
 
 func TestMaxIdleTimeStopsTheReflector(t *testing.T) {
+	tCtx := ktesting.Init(t)
 	secret := &v1.Secret{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:            "name",
@@ -364,7 +365,7 @@ func TestMaxIdleTimeStopsTheReflector(t *testing.T) {
 	fakeWatch := watch.NewFake()
 	fakeClient.AddWatchReactor("secrets", core.DefaultWatchReactor(fakeWatch, nil))
 	fakeClock := testingclock.NewFakeClock(time.Now())
-	store := newSecretCache(fakeClient, fakeClock, time.Minute)
+	store := newSecretCache(tCtx, fakeClient, fakeClock, time.Minute)
 
 	key := objectKey{namespace: "ns", name: "name"}
 	itemExists := func(_ context.Context) (bool, error) {
@@ -386,7 +387,6 @@ func TestMaxIdleTimeStopsTheReflector(t *testing.T) {
 
 	// AddReference should start reflector.
 	store.AddReference("ns", "name", "pod")
-	tCtx := ktesting.Init(t)
 	if err := wait.PollUntilContextCancel(tCtx, 10*time.Millisecond, false, itemExists); err != nil {
 		t.Errorf("item wasn't added to cache")
 	}
@@ -420,6 +420,7 @@ func TestMaxIdleTimeStopsTheReflector(t *testing.T) {
 }
 
 func TestReflectorNotStoppedOnSlowInitialization(t *testing.T) {
+	tCtx := ktesting.Init(t)
 	secret := &v1.Secret{
 		ObjectMeta: metav1.ObjectMeta{
 			Name:            "name",
@@ -447,7 +448,7 @@ func TestReflectorNotStoppedOnSlowInitialization(t *testing.T) {
 	fakeClient.AddReactor("list", "secrets", listReactor)
 	fakeWatch := watch.NewFake()
 	fakeClient.AddWatchReactor("secrets", core.DefaultWatchReactor(fakeWatch, nil))
-	store := newSecretCache(fakeClient, fakeClock, time.Minute)
+	store := newSecretCache(tCtx, fakeClient, fakeClock, time.Minute)
 
 	key := objectKey{namespace: "ns", name: "name"}
 	itemExists := func(_ context.Context) (bool, error) {
@@ -479,7 +480,6 @@ func TestReflectorNotStoppedOnSlowInitialization(t *testing.T) {
 
 	// AddReference should start reflector.
 	store.AddReference("ns", "name", "pod")
-	tCtx := ktesting.Init(t)
 	if err := wait.PollUntilContextCancel(tCtx, 10*time.Millisecond, false, itemExists); err != nil {
 		t.Errorf("item wasn't added to cache")
 	}
@@ -592,6 +592,7 @@ func TestRefMapHandlesReferencesCorrectly(t *testing.T) {
 	}
 	for _, tc := range tests {
 		t.Run(tc.desc, func(t *testing.T) {
+			tCtx := ktesting.Init(t)
 			fakeClient := &fake.Clientset{}
 			listReactor := func(a core.Action) (bool, runtime.Object, error) {
 				result := &v1.SecretList{
@@ -606,7 +607,7 @@ func TestRefMapHandlesReferencesCorrectly(t *testing.T) {
 			fakeWatch := watch.NewFake()
 			fakeClient.AddWatchReactor("secrets", core.DefaultWatchReactor(fakeWatch, nil))
 			fakeClock := testingclock.NewFakeClock(time.Now())
-			store := newSecretCache(fakeClient, fakeClock, time.Minute)
+			store := newSecretCache(tCtx, fakeClient, fakeClock, time.Minute)
 
 			for i, step := range tc.steps {
 				expect := tc.expects[i]

--- a/pkg/kubelet/util/manager/watch_based_manager_test.go
+++ b/pkg/kubelet/util/manager/watch_based_manager_test.go
@@ -275,7 +275,7 @@ func TestImmutableSecretStopsTheReflector(t *testing.T) {
 			fakeClient.AddWatchReactor("secrets", core.DefaultWatchReactor(fakeWatch, nil))
 
 			fakeClock := testingclock.NewFakeClock(time.Now())
-			store := newSecretCache(fakeClient, fakeClock, time.Minute)
+			store := newSecretCache(tCtx, fakeClient, fakeClock, time.Minute)
 
 			key := objectKey{namespace: "ns", name: "name"}
 			itemExists := func(_ context.Context) (bool, error) {

--- a/pkg/kubelet/util/nodelease.go
+++ b/pkg/kubelet/util/nodelease.go
@@ -29,14 +29,14 @@ import (
 
 // SetNodeOwnerFunc helps construct a newLeasePostProcessFunc which sets
 // a node OwnerReference to the given lease object
-func SetNodeOwnerFunc(c clientset.Interface, nodeName string) func(lease *coordinationv1.Lease) error {
+func SetNodeOwnerFunc(ctx context.Context, c clientset.Interface, nodeName string) func(lease *coordinationv1.Lease) error {
 	return func(lease *coordinationv1.Lease) error {
 		// Setting owner reference needs node's UID. Note that it is different from
 		// kubelet.nodeRef.UID. When lease is initially created, it is possible that
 		// the connection between master and node is not ready yet. So try to set
 		// owner reference every time when renewing the lease, until successful.
 		if len(lease.OwnerReferences) == 0 {
-			if node, err := c.CoreV1().Nodes().Get(context.TODO(), nodeName, metav1.GetOptions{}); err == nil {
+			if node, err := c.CoreV1().Nodes().Get(ctx, nodeName, metav1.GetOptions{}); err == nil {
 				lease.OwnerReferences = []metav1.OwnerReference{
 					{
 						APIVersion: corev1.SchemeGroupVersion.WithKind("Node").Version,
@@ -46,7 +46,8 @@ func SetNodeOwnerFunc(c clientset.Interface, nodeName string) func(lease *coordi
 					},
 				}
 			} else {
-				klog.ErrorS(err, "Failed to get node when trying to set owner ref to the node lease", "node", klog.KRef("", nodeName))
+				logger := klog.FromContext(ctx)
+				logger.Error(err, "Failed to get node when trying to set owner ref to the node lease", "node", klog.KRef("", nodeName))
 				return err
 			}
 		}

--- a/pkg/kubelet/util/pod_startup_latency_tracker.go
+++ b/pkg/kubelet/util/pod_startup_latency_tracker.go
@@ -17,6 +17,7 @@ limitations under the License.
 package util
 
 import (
+	"context"
 	"sync"
 	"time"
 
@@ -97,11 +98,14 @@ func (p *basicPodStartupLatencyTracker) ObservedPodOnWatch(pod *v1.Pod, when tim
 	}
 
 	if hasPodStartedSLO(pod) {
+		// TODO: it needs to be replaced by a proper context in the future
+		ctx := context.TODO()
+		logger := klog.FromContext(ctx)
 		podStartingDuration := when.Sub(pod.CreationTimestamp.Time)
 		imagePullingDuration := state.lastFinishedPulling.Sub(state.firstStartedPulling)
 		podStartSLOduration := (podStartingDuration - imagePullingDuration).Seconds()
 
-		klog.InfoS("Observed pod startup duration",
+		logger.Info("Observed pod startup duration",
 			"pod", klog.KObj(pod),
 			"podStartSLOduration", podStartSLOduration,
 			"podStartE2EDuration", podStartingDuration,
@@ -168,8 +172,11 @@ func (p *basicPodStartupLatencyTracker) RecordStatusUpdated(pod *v1.Pod) {
 		return
 	}
 
+	// TODO: it needs to be replaced by a proper context in the future
+	ctx := context.TODO()
+	logger := klog.FromContext(ctx)
 	if hasPodStartedSLO(pod) {
-		klog.V(3).InfoS("Mark when the pod was running for the first time", "pod", klog.KObj(pod), "rv", pod.ResourceVersion)
+		logger.V(3).Info("Mark when the pod was running for the first time", "pod", klog.KObj(pod), "rv", pod.ResourceVersion)
 		state.observedRunningTime = p.clock.Now()
 	}
 }

--- a/pkg/kubelet/util/swap/swap_util.go
+++ b/pkg/kubelet/util/swap/swap_util.go
@@ -18,6 +18,7 @@ package swap
 
 import (
 	"bytes"
+	"context"
 	"errors"
 	"os"
 	sysruntime "runtime"
@@ -42,6 +43,9 @@ var (
 const TmpfsNoswapOption = "noswap"
 
 func IsTmpfsNoswapOptionSupported(mounter mount.Interface, mountPath string) bool {
+	// TODO: it needs to be replaced by a proper context in the future
+	ctx := context.TODO()
+	logger := klog.FromContext(ctx)
 	isTmpfsNoswapOptionSupportedHelper := func() bool {
 		if sysruntime.GOOS == "windows" {
 			return false
@@ -51,13 +55,13 @@ func IsTmpfsNoswapOptionSupported(mounter mount.Interface, mountPath string) boo
 			// Turning off swap in unprivileged tmpfs mounts unsupported
 			// https://github.com/torvalds/linux/blob/v6.8/mm/shmem.c#L4004-L4011
 			// https://github.com/kubernetes/kubernetes/issues/125137
-			klog.InfoS("Running under a user namespace - tmpfs noswap is not supported")
+			logger.Info("Running under a user namespace - tmpfs noswap is not supported")
 			return false
 		}
 
 		kernelVersion, err := utilkernel.GetVersion()
 		if err != nil {
-			klog.ErrorS(err, "cannot determine kernel version, unable to determine is tmpfs noswap is supported")
+			logger.Error(err, "cannot determine kernel version, unable to determine is tmpfs noswap is supported")
 			return false
 		}
 
@@ -66,31 +70,31 @@ func IsTmpfsNoswapOptionSupported(mounter mount.Interface, mountPath string) boo
 		}
 
 		if mountPath == "" {
-			klog.ErrorS(errors.New("mount path is empty, falling back to /tmp"), "")
+			logger.Error(errors.New("mount path is empty, falling back to /tmp"), "")
 		}
 
 		mountPath, err = os.MkdirTemp(mountPath, "tmpfs-noswap-test-")
 		if err != nil {
-			klog.InfoS("error creating dir to test if tmpfs noswap is enabled. Assuming not supported", "mount path", mountPath, "error", err)
+			logger.Info("error creating dir to test if tmpfs noswap is enabled. Assuming not supported", "mount path", mountPath, "error", err)
 			return false
 		}
 
 		defer func() {
 			err = os.RemoveAll(mountPath)
 			if err != nil {
-				klog.ErrorS(err, "error removing test tmpfs dir", "mount path", mountPath)
+				logger.Error(err, "error removing test tmpfs dir", "mount path", mountPath)
 			}
 		}()
 
 		err = mounter.MountSensitiveWithoutSystemd("tmpfs", mountPath, "tmpfs", []string{TmpfsNoswapOption}, nil)
 		if err != nil {
-			klog.InfoS("error mounting tmpfs with the noswap option. Assuming not supported", "error", err)
+			logger.Info("error mounting tmpfs with the noswap option. Assuming not supported", "error", err)
 			return false
 		}
 
 		err = mounter.Unmount(mountPath)
 		if err != nil {
-			klog.ErrorS(err, "error unmounting test tmpfs dir", "mount path", mountPath)
+			logger.Error(err, "error unmounting test tmpfs dir", "mount path", mountPath)
 		}
 
 		return true
@@ -104,7 +108,7 @@ func IsTmpfsNoswapOptionSupported(mounter mount.Interface, mountPath string) boo
 }
 
 // gets /proc/swaps's content as an input, returns true if swap is enabled.
-func isSwapOnAccordingToProcSwaps(procSwapsContent []byte) bool {
+func isSwapOnAccordingToProcSwaps(logger klog.Logger, procSwapsContent []byte) bool {
 	procSwapsContent = bytes.TrimSpace(procSwapsContent) // extra trailing \n
 	procSwapsStr := string(procSwapsContent)
 	procSwapsLines := strings.Split(procSwapsStr, "\n")
@@ -112,7 +116,7 @@ func isSwapOnAccordingToProcSwaps(procSwapsContent []byte) bool {
 	// If there is more than one line (table headers) in /proc/swaps then swap is enabled
 	isSwapOn := len(procSwapsLines) > 1
 	if isSwapOn {
-		klog.InfoS("Swap is on", "/proc/swaps contents", procSwapsStr)
+		logger.Info("Swap is on", "/proc/swaps contents", procSwapsStr)
 	}
 
 	return isSwapOn
@@ -127,18 +131,22 @@ func IsSwapOn() (bool, error) {
 			return false, nil
 		}
 
+		// TODO: it needs to be replaced by a proper context in the future
+		ctx := context.TODO()
+		logger := klog.FromContext(ctx)
+
 		const swapFilePath = "/proc/swaps"
 		procSwapsContent, err := os.ReadFile(swapFilePath)
 		if err != nil {
 			if os.IsNotExist(err) {
-				klog.InfoS("File does not exist, assuming that swap is disabled", "path", swapFilePath)
+				logger.Info("File does not exist, assuming that swap is disabled", "path", swapFilePath)
 				return false, nil
 			}
 
 			return false, err
 		}
 
-		return isSwapOnAccordingToProcSwaps(procSwapsContent), nil
+		return isSwapOnAccordingToProcSwaps(logger, procSwapsContent), nil
 	}
 
 	swapOnOnce.Do(func() {

--- a/pkg/kubelet/util/swap/swap_util_test.go
+++ b/pkg/kubelet/util/swap/swap_util_test.go
@@ -16,7 +16,12 @@ limitations under the License.
 
 package swap
 
-import "testing"
+import (
+	"testing"
+
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/test/utils/ktesting"
+)
 
 func TestIsSwapEnabled(t *testing.T) {
 	testCases := []struct {
@@ -54,10 +59,12 @@ Filename				Type		Size		Used		Priority
 			expectedEnabled: false,
 		},
 	}
+	tCtx := ktesting.Init(t)
+	logger := klog.FromContext(tCtx)
 
 	for _, tc := range testCases {
 		t.Run(tc.name, func(t *testing.T) {
-			isEnabled := isSwapOnAccordingToProcSwaps([]byte(tc.procSwapsContent))
+			isEnabled := isSwapOnAccordingToProcSwaps(logger, []byte(tc.procSwapsContent))
 			if isEnabled != tc.expectedEnabled {
 				t.Errorf("expected %v, got %v", tc.expectedEnabled, isEnabled)
 			}

